src/main/java/games/strategy/engine/lobby/client/ui/TimespanDialog.java

```diff
@@ -0,0 +1,107 @@
+package games.strategy.engine.lobby.client.ui;
+
+import java.awt.Component;
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.time.temporal.ChronoUnit;
+import java.util.Date;
+import java.util.function.Consumer;
+import java.util.function.Function;
+
+import javax.swing.JComboBox;
+import javax.swing.JOptionPane;
+import javax.swing.JSpinner;
+import javax.swing.SpinnerNumberModel;
+import javax.swing.border.EmptyBorder;
+import javax.swing.event.AncestorEvent;
+import javax.swing.event.AncestorListener;
+
+import com.google.common.annotations.VisibleForTesting;
+
+import swinglib.JLabelBuilder;
+import swinglib.JPanelBuilder;
+
+/**
+ * A UI-Utility class that can be used to prompt the user for a ban or mute time.
+ */
+public class TimespanDialog {
+
+  /**
+   * The possible time units and corresponding mappings.
+   */
+  @VisibleForTesting
+  enum TimeUnit {
+    MINUTES("Minutes", i -> Instant.now().plus(i, ChronoUnit.MINUTES)),
+    HOURS("Hours", i -> Instant.now().plus(i, ChronoUnit.HOURS)),
+    DAYS("Days", i -> Instant.now().plus(i, ChronoUnit.DAYS)),
+    WEEKS("Weeks", i -> LocalDateTime.now(ZoneOffset.UTC).plus(i, ChronoUnit.WEEKS).toInstant(ZoneOffset.UTC)),
+    MONTHS("Months", i -> LocalDateTime.now(ZoneOffset.UTC).plus(i, ChronoUnit.MONTHS).toInstant(ZoneOffset.UTC)),
+    YEARS("Years", i -> LocalDateTime.now(ZoneOffset.UTC).plus(i, ChronoUnit.YEARS).toInstant(ZoneOffset.UTC)),
+    FOREVER("Forever", i -> null);
+
+    private final String name;
+    private final Function<Integer, Instant> function;
+
+    private TimeUnit(final String name, final Function<Integer, Instant> function) {
+      this.name = name;
+      this.function = function;
+    }
+
+    @Override
+    public String toString() {
```

Comment:
Minor, I'm a bit confused why a `TimespanDialog` has a `toString` function,
Trying to answer that, I notice that the `name` value is passed in via constructor and then we simply return that. For some reason then this dialog is also a 'carrier' for this name value and is independent of everyting else in this class (unless I missed a usage somewhere which is very possible).

I'll try to come back to this PR tomorrow to give a more detailed feedback, apologies for not quite fully tracing out things to answer my own questions.

Reply:
Ah, it's the toString() of the enum type... that helps.
At this point I would suggest we rename the `name` variable, it's not very useful for knowing what we are naming : )

Reply:
The  toString method is veing invoked by the default cellrenderer of JComboBox.
Either the toString methods gets overriden (simpler), or we create our own CellRenderer to rendem our enum customly, in order to have a normal case (no caps)

Reply:
I see, thanks.
I wonder why we are allowing just this much flexibility. Who really needs to specify number of years when you can just say 365 days? I would recommend long term we simplify this and create a code enforced ban policy.

First ban = 15 minutes, second = 1 day, third = 1 week, fourth = permanent.

Then mods will just click ban and the code will do the right thing. Along the way we also record these bans so we don't have the overhead of mods to report their actions.

Given all that, for what we have, the enum is actually pretty slick/reasonable, but I think the premise of what we are doing is a bit misguided with too much flexibility allowed (and hence all these warnings about asking other mods before doing more than 1 day bans, and also unclear how we can have a consistent ban policy if we are arbitrarily picking hours or days)